spring-cloud-gateway-server/src/main/java/org/springframework/cloud/gateway/filter/factory/JsonToGrpcGatewayFilterFactory.java

```diff
@@ -0,0 +1,308 @@
+
+/**
+ * This filter takes a JSON payload, transform it into a protobuf object, send it to a
+ * given gRPC channel, and transform the response back to JSON.
+ *
+ * Making it transparent for the consumer that the service under the gateway is a gRPC
+ * one.
+ *
+ * @author Alberto C. Ríos
+ */
+public class JsonToGrpcGatewayFilterFactory
+		extends AbstractGatewayFilterFactory<JsonToGrpcGatewayFilterFactory.Config> {
+
+	private final GRPCSSLContext sslContext;
+
+	public JsonToGrpcGatewayFilterFactory(GRPCSSLContext sslContext) {
+		super(Config.class);
+		this.sslContext = sslContext;
+	}
+
+	@Override
+	public List<String> shortcutFieldOrder() {
+		return Arrays.asList("protoDescriptor", "protoFile", "service", "method");
+	}
+
+	@Override
+	public GatewayFilter apply(Config config) {
+		GatewayFilter filter = new GatewayFilter() {
+			@Override
+			public Mono<Void> filter(ServerWebExchange exchange, GatewayFilterChain chain) {
+				GRPCResponseDecorator modifiedResponse = new GRPCResponseDecorator(exchange, config);
+
+				return modifiedResponse.writeWith(exchange.getRequest().getBody())
+						.then(chain.filter(exchange.mutate().response(modifiedResponse).build()));
+			}
+
+			@Override
+			public String toString() {
+				return filterToStringCreator(JsonToGrpcGatewayFilterFactory.this).toString();
+			}
+		};
+
+		int order = NettyWriteResponseFilter.WRITE_RESPONSE_FILTER_ORDER - 1;
+		return new OrderedGatewayFilter(filter, order);
+	}
+
+	public static class Config {
+
+		private String protoDescriptor;
+
+		private String protoFile;
+
+		private String service;
+
+		private String method;
+
+		public String getProtoDescriptor() {
+			return protoDescriptor;
+		}
+
+		public Config setProtoDescriptor(String protoDescriptor) {
+			this.protoDescriptor = protoDescriptor;
+			return this;
+		}
+
+		public String getProtoFile() {
+			return protoFile;
+		}
+
+		public Config setProtoFile(String protoFile) {
+			this.protoFile = protoFile;
+			return this;
+		}
+
+		public String getService() {
+			return service;
+		}
+
+		public Config setService(String service) {
+			this.service = service;
+			return this;
+		}
+
+		public String getMethod() {
+			return method;
+		}
+
+		public Config setMethod(String method) {
+			this.method = method;
+			return this;
+		}
+
+	}
+
+	class GRPCResponseDecorator extends ServerHttpResponseDecorator {
+
+		private final ServerWebExchange exchange;
+
+		private final Descriptors.Descriptor descriptor;
+
+		private final ObjectWriter objectWriter;
+
+		private final ObjectReader objectReader;
+
+		private final ClientCall<DynamicMessage, DynamicMessage> clientCall;
+
+		private final ObjectNode objectNode;
+
+		GRPCResponseDecorator(ServerWebExchange exchange, Config config) {
+			super(exchange.getResponse());
+			this.exchange = exchange;
+			try {
+				File descriptorFile = new File(config.getProtoDescriptor());
```

Comment:
Let's treat these as spring [resources](https://docs.spring.io/spring-framework/docs/current/reference/html/core.html#resources). That way there's more flexible resolving rather than just filesystem, such as classpath. Inject a `ResourceLoader` into the factory and then here call `ResourceLoader.getResource()`.

Reply:
Excellent idea! Much more flexible than being tied to a file resource.